# Lecture 5: Guided Examples 3.14.1–3.14.8

The exact example and exercise numbers match the live tutorial. Open an item from the index below, or search for its full number. Run **Setup** before the core examples and solutions. Tutorial 3.14 contains standalone guided examples. The data are synthetic teaching records.


## Numbered index

| Number | Exact tutorial title |
|---|---|
| [Example 3.14.1](#scrollTo=l05-example-3-14-1) | Identify and describe the dataset |
| [Example 3.14.2](#scrollTo=l05-example-3-14-2) | Inspect the structure and sample records |
| [Example 3.14.3](#scrollTo=l05-example-3-14-3) | Calculate preliminary numerical summaries |
| [Example 3.14.4](#scrollTo=l05-example-3-14-4) | Create and inspect derived measures |
| [Example 3.14.5](#scrollTo=l05-example-3-14-5) | Explore categories and group differences |
| [Example 3.14.6](#scrollTo=l05-example-3-14-6) | Complete a structured data-quality audit |
| [Example 3.14.7](#scrollTo=l05-example-3-14-7) | Explore the feedback qualitatively and computationally |
| [Example 3.14.8](#scrollTo=l05-example-3-14-8) | Write the Initial Data Exploration Report |


## Setup

Run this cell first. It loads the versioned raw CSV from GitHub and checks the expected 121 × 14 structure. Keep the raw data unchanged during Lecture 5.


In [1]:
import pandas as pd  # Loads the library feature used by this step.
data_url = "https://raw.githubusercontent.com/asmrabbi/E26_TAN7_Scripting_CPH/main/data/E26_TAN7_service_experience_raw.csv"  # Stores the calculated or supplied value for later use.
df = pd.read_csv(data_url)  # Stores the calculated or supplied value for later use.
print("Dataset shape:", df.shape)  # Displays the result so it can be inspected.
assert df.shape == (121, 14), "The wrong dataset version was loaded."  # Stores the calculated or supplied value for later use.
display(df.head())  # Displays the result so it can be inspected.


Dataset shape: (121, 14)
     record_id report_month        city      service_type  cases_received  cases_resolved resolution_days  satisfaction_score  digital_cases  in_person_cases  repeat_contacts  complaints_received  staff_hours                                                                         feedback
0  SR2026-0001   2026-01-01  Copenhagen           Housing             130             116             8.0                 3.5             88               42               17                   10         87.8  The guidance was clear, but the supporting-document requirements caused delays.
1  SR2026-0002   2026-01-01  Copenhagen         Transport             124             110             5.1                 3.9             92               32               10                    8         54.7               Several users asked for clearer information about service changes.
2  SR2026-0003   2026-01-01  Copenhagen        Employment             105              93             5.5

# Tutorial 3.14


## Example 3.14.1 — Identify and describe the dataset

[Open Example 3.14.1 on the website](https://asmrabbi.github.io/E26_TAN7_Scripting_CPH/#data-i/3.14/example-3.14.1)


### Question and goal

Write a short dataset identity section that states:

1. dataset name;
2. source/location;
3. whether the data are real or synthetic;
4. intended unit of observation;
5. number of rows and columns;
6. all column names;
7. which columns appear to be identifiers, categorical, numeric, date-like and free text.

### Required code

At minimum use:


In [2]:
df.shape
df.columns
df.dtypes
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121 entries, 0 to 120
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   record_id            121 non-null    object 
 1   report_month         121 non-null    object 
 2   city                 121 non-null    object 
 3   service_type         121 non-null    object 
 4   cases_received       121 non-null    int64  
 5   cases_resolved       121 non-null    int64  
 6   resolution_days      121 non-null    object 
 7   satisfaction_score   117 non-null    float64
 8   digital_cases        121 non-null    int64  
 9   in_person_cases      121 non-null    int64  
 10  repeat_contacts      121 non-null    int64  
 11  complaints_received  121 non-null    int64  
 12  staff_hours          121 non-null    float64
 13  feedback             118 non-null    object 
dtypes: float64(2), int64(6), object(6)
memory usage: 13.4+ KB


### Required written output

Write 4–6 sentences. Do not simply paste the code output.

<details>
<summary>What the result means</summary>

The dataset is `E26_TAN7_service_experience_raw.csv`, a synthetic teaching file stored in the E26 course GitHub repository. One row is intended to represent one service category in one city during one reporting month. The raw DataFrame contains 121 rows and 14 columns. `record_id` functions conceptually as an identifier; `city` and `service_type` are categorical; seven columns are intended as counts or measures; `report_month` is date-like; and `feedback` is free text. The pandas dtypes do not perfectly match the conceptual schema because date/numeric-like values are initially stored as text-like data and recognised missingness occurs in two columns. The file is designed for teaching and must not be interpreted as evidence about actual municipal service performance.

</details>

---


## Example 3.14.2 — Inspect the structure and sample records

[Open Example 3.14.2 on the website](https://asmrabbi.github.io/E26_TAN7_Scripting_CPH/#data-i/3.14/example-3.14.2)


### Question and goal

Inspect:

- first five rows;
- final five rows;
- five sampled rows;
- shape;
- non-missing counts;
- dtypes.

Use a reproducible sample:


In [3]:
df.sample(5, random_state=42)


      record_id report_month        city      service_type  cases_received  cases_resolved resolution_days  satisfaction_score  digital_cases  in_person_cases  repeat_contacts  complaints_received  staff_hours                                                                               feedback
44  SR2026-0045   2026-03-01  Copenhagen  Citizen Services             185             174             2.6                 4.2            154               31               10                    6         56.2                   The online process was easy to use, although one step was confusing.
47  SR2026-0048   2026-03-01      Aarhus        Employment              96              90             4.4                 3.9             70               26               12                    5         55.1  The online information was helpful, but some users wanted more detailed instructions.
4   SR2026-0005   2026-01-01  Copenhagen  Citizen Services             166             152             3.1   

Then answer:

1. Which columns look straightforwardly numeric?
2. Which columns need closer inspection before numeric/date analysis?
3. Why is inspecting only `head()` insufficient?

### Complete worked code


In [4]:
print("Shape:", df.shape)

display(df.head())
display(df.tail())
display(df.sample(5, random_state=42))

print(df.dtypes)
df.info()


Shape: (121, 14)
     record_id report_month        city      service_type  cases_received  cases_resolved resolution_days  satisfaction_score  digital_cases  in_person_cases  repeat_contacts  complaints_received  staff_hours                                                                         feedback
0  SR2026-0001   2026-01-01  Copenhagen           Housing             130             116             8.0                 3.5             88               42               17                   10         87.8  The guidance was clear, but the supporting-document requirements caused delays.
1  SR2026-0002   2026-01-01  Copenhagen         Transport             124             110             5.1                 3.9             92               32               10                    8         54.7               Several users asked for clearer information about service changes.
2  SR2026-0003   2026-01-01  Copenhagen        Employment             105              93             5.5        

---


## Example 3.14.3 — Calculate preliminary numerical summaries

[Open Example 3.14.3 on the website](https://asmrabbi.github.io/E26_TAN7_Scripting_CPH/#data-i/3.14/example-3.14.3)


### Question and goal

For `cases_received`, calculate:

- count;
- sum;
- mean;
- median;
- standard deviation;
- minimum;
- maximum;
- `describe()`.

For `cases_resolved`, calculate:

- sum;
- mean;
- median.

For `satisfaction_score`, calculate:

- non-missing count;
- mean;
- median;
- minimum;
- maximum.

### Complete worked code


In [5]:
print("Cases received")
print("Count:", df["cases_received"].count())
print("Sum:", df["cases_received"].sum())
print("Mean:", df["cases_received"].mean())
print("Median:", df["cases_received"].median())
print("Std:", df["cases_received"].std())
print("Min:", df["cases_received"].min())
print("Max:", df["cases_received"].max())

print("\nFull summary")
print(df["cases_received"].describe())


Cases received
Count: 121
Sum: 12653
Mean: 104.5702479338843
Median: 102.0
Std: 27.256750370712755
Min: -4
Max: 185

Full summary
count    121.000000
mean     104.570248
std       27.256750
min       -4.000000
25%       86.000000
50%      102.000000
75%      122.000000
max      185.000000
Name: cases_received, dtype: float64


### Interpretation questions

1. Which result immediately suggests a possible validity problem?
2. Why should the raw mean not yet be described as a final result?
3. How many satisfaction scores contribute to the raw mean?
4. Does the raw satisfaction maximum fit the documented 1–5 scale?

<details>
<summary>What the result means</summary>

For raw `cases_received`:

```text
count = 121
sum = 12653
mean ≈ 104.57
median = 102.0
std ≈ 27.26
min = -4
max = 185
```

The negative minimum is a clear quality flag under the current schema.

For raw satisfaction:

```text
non-missing count = 117
mean ≈ 4.11
min = 0.0
max = 6.2
```

The maximum conflicts with the documented 1–5 range, so the mean should remain preliminary.

</details>

---


## Example 3.14.4 — Create and inspect derived measures

[Open Example 3.14.4 on the website](https://asmrabbi.github.io/E26_TAN7_Scripting_CPH/#data-i/3.14/example-3.14.4)


### Question and goal

Create:

```text
unresolved_cases = cases_received - cases_resolved
```

and:

```text
resolution_rate = cases_resolved / cases_received × 100
```

Then inspect:

- negative unresolved values;
- rates above 100%;
- missing/undefined rates caused by a zero denominator.

### Complete worked code


In [6]:
df["unresolved_cases"] = (
    df["cases_received"] - df["cases_resolved"]
)

df["resolution_rate"] = (
    df["cases_resolved"] /
    df["cases_received"] * 100
)

columns = [
    "record_id",
    "cases_received",
    "cases_resolved",
    "unresolved_cases",
    "resolution_rate"
]

display(df[columns])


      record_id  cases_received  cases_resolved  unresolved_cases  resolution_rate
0   SR2026-0001             130             116                14        89.230769
1   SR2026-0002             124             110                14        88.709677
2   SR2026-0003             105              93                12        88.571429
3   SR2026-0004             104              94                10        90.384615
4   SR2026-0005             166             152                14        91.566265
5   SR2026-0006             111              96                15        86.486486
6   SR2026-0007              95              88                 7        92.631579
7   SR2026-0008              87              80                 7        91.954023
8   SR2026-0009              97              89                 8        91.752577
9   SR2026-0010             123             113                10        91.869919
10  SR2026-0011             107              92                15        85.981308
11  

### Additional calculations

Calculate both:


In [7]:
mean_row_rate = df["resolution_rate"].mean()


and:


In [8]:
overall_rate = (
    df["cases_resolved"].sum() /
    df["cases_received"].sum() * 100
)


### Questions

1. Why do these two results differ?
2. Which denominator does each method use conceptually?
3. Why should neither raw result yet be treated as final?

<details>
<summary>What the result means</summary>

The mean row rate gives each row equal weight, while the aggregate rate divides the total number resolved by the total number received and therefore weights records through their case volumes. In the supplied raw teaching data they are approximately `93.66%` and `94.51%`, respectively. The calculations are affected by problematic raw values, including three rows with resolved cases greater than received cases and one negative received count. The current file has no zero denominator, but the method still requires a documented zero-case rule. The correct analytical measure must therefore be defined and the source data validated before reporting a final rate.

</details>

---


## Example 3.14.5 — Explore categories and group differences

[Open Example 3.14.5 on the website](https://asmrabbi.github.io/E26_TAN7_Scripting_CPH/#data-i/3.14/example-3.14.5)


### Question and goal

For `service_type`:

1. print unique values;
2. calculate counts;
3. calculate percentages;
4. calculate mean and median `cases_received` by service type;
5. calculate count, mean, median, minimum and maximum `satisfaction_score` by service type.

For `city`:

1. print unique raw labels;
2. calculate raw frequencies;
3. calculate grouped mean `cases_received` by raw city label;
4. explain why the result should not yet be interpreted as a clean city comparison.

Create a crosstab of:

```text
city × service_type
```

### Complete worked code


In [9]:
print(df["service_type"].value_counts())
print(df["service_type"].value_counts(normalize=True).mul(100).round(2))

service_cases = (
    df.groupby("service_type")["cases_received"]
      .agg(["count", "sum", "mean", "median", "min", "max"])
)

display(service_cases.round(2))

service_satisfaction = (
    df.groupby("service_type")["satisfaction_score"]
      .agg(["count", "mean", "median", "min", "max"])
)

display(service_satisfaction.round(2))

print(df["city"].unique())
print(df["city"].value_counts())

city_service = pd.crosstab(
    df["city"],
    df["service_type"]
)

display(city_service)


service_type
Transport              24
Employment             24
Waste                  24
Citizen Services       24
Housing                21
Transport               1
Citizen services        1
WASTE                   1
Employment Services     1
Name: count, dtype: int64
service_type
Transport              19.83
Employment             19.83
Waste                  19.83
Citizen Services       19.83
Housing                17.36
Transport               0.83
Citizen services        0.83
WASTE                   0.83
Employment Services     0.83
Name: proportion, dtype: float64
                     count   sum    mean  median  min  max
service_type                                              
Citizen Services        24  3118  129.92   122.0   94  185
Citizen services         1    99   99.00    99.0   99   99
Employment              24  2113   88.04    86.5   62  129
Employment Services      1    97   97.00    97.0   97   97
Housing                 21  2433  115.86   113.0   86  157
Transpo

### Required interpretation

Write at least two preliminary observations and then explicitly state which raw-data problems could distort them.

<details>
<summary>What the result means</summary>

The five most frequent exact raw service labels are:

```text
Citizen Services    24
Employment          24
Transport           24
Waste               24
Housing             21
```

Their raw proportions are approximately 19.83% for each 24-row label and 17.36% for Housing. Four inconsistent one-row service labels each account for 0.83%.

Raw mean `cases_received` for the five main exact labels is approximately:

```text
Citizen Services    129.92
Employment           88.04
Housing              115.86
Transport            102.54
Waste                 88.96
```

However, Waste includes the negative raw count, the dataset contains an exact duplicate record, and four inconsistent service labels are split into separate groups. City-level grouping is also problematic because several raw strings appear to represent inconsistent versions of city names.

</details>

---


## Example 3.14.6 — Complete a structured data-quality audit

[Open Example 3.14.6 on the website](https://asmrabbi.github.io/E26_TAN7_Scripting_CPH/#data-i/3.14/example-3.14.6)


### Question and goal

Create checks for:

1. missing values by column;
2. missing percentages;
3. exact duplicate rows;
4. repeated `record_id` values;
5. raw city-category inconsistencies;
6. negative `cases_received`;
7. satisfaction outside 1–5;
8. `cases_resolved > cases_received`;
9. non-numeric `resolution_days`;
10. invalid `report_month` values.

### Complete worked code


In [10]:
print("Missing values")
print(df.isna().sum())

print("\nMissing percentages")
print((df.isna().mean() * 100).round(2))

print("\nExact duplicates")
print(df.duplicated().sum())
display(df[df.duplicated(keep=False)])

print("\nRepeated IDs")
id_duplicate_mask = df["record_id"].duplicated(keep=False)
display(df[id_duplicate_mask])

print("\nRaw city labels")
print(df["city"].value_counts(dropna=False))

print("\nNegative received cases")
display(df[df["cases_received"] < 0])

print("\nOut-of-range satisfaction")
sat_problem = (
    (df["satisfaction_score"] < 1) |
    (df["satisfaction_score"] > 5)
)
display(df[sat_problem])

print("\nResolved greater than received")
display(df[df["cases_resolved"] > df["cases_received"]])

resolution_preview = pd.to_numeric(
    df["resolution_days"],
    errors="coerce"
)
resolution_problem = (
    df["resolution_days"].notna() &
    resolution_preview.isna()
)
print("\nNon-numeric resolution values")
display(df.loc[resolution_problem, ["record_id", "resolution_days"]])

month_preview = pd.to_datetime(
    df["report_month"],
    errors="coerce"
)
print("\nInvalid report months")
display(df.loc[month_preview.isna(), ["record_id", "report_month"]])


Missing values
record_id              0
report_month           0
city                   0
service_type           0
cases_received         0
cases_resolved         0
resolution_days        0
satisfaction_score     4
digital_cases          0
in_person_cases        0
repeat_contacts        0
complaints_received    0
staff_hours            0
feedback               3
unresolved_cases       0
resolution_rate        0
dtype: int64

Missing percentages
record_id              0.00
report_month           0.00
city                   0.00
service_type           0.00
cases_received         0.00
cases_resolved         0.00
resolution_days        0.00
satisfaction_score     3.31
digital_cases          0.00
in_person_cases        0.00
repeat_contacts        0.00
complaints_received    0.00
staff_hours            0.00
feedback               2.48
unresolved_cases       0.00
resolution_rate        0.00
dtype: float64

Exact duplicates
1
       record_id report_month    city service_type  cases_received  

### Required output: quality log

Create a Markdown table with at least these columns:

| Issue | How detected | Affected variable/record | Why it matters | What should happen next? |
|---|---|---|---|---|

Do not put “delete” or “replace” as an automatic next step unless you first explain the evidence required to justify that action.

---


## Example 3.14.7 — Explore the feedback qualitatively and computationally

[Open Example 3.14.7 on the website](https://asmrabbi.github.io/E26_TAN7_Scripting_CPH/#data-i/3.14/example-3.14.7)


### Question and goal

1. Display all non-missing feedback.
2. Read the comments manually.
3. Identify at least three recurring ideas.
4. Create at least two transparent Boolean keyword indicators.
5. Count each indicator.
6. Display the matching comments.
7. Compare one indicator with `service_type` or `satisfaction_score`.
8. Describe at least one false-positive/false-negative risk.

### Example indicator setup


In [11]:
df["mentions_wait"] = df["feedback"].str.contains(
    "wait|long",
    case=False,
    na=False,
    regex=True
)

df["mentions_helpful"] = df["feedback"].str.contains(
    "helpful",
    case=False,
    na=False
)


### Count the rules


In [12]:
print("Waiting matches:", df["mentions_wait"].sum())
print("Helpful matches:", df["mentions_helpful"].sum())


Waiting matches: 11
Helpful matches: 36


Expected raw counts:

```text
Waiting matches: 11
Helpful matches: 36
```

### Compare waiting indicator and service type


In [13]:
wait_by_service = pd.crosstab(
    df["service_type"],
    df["mentions_wait"]
)

display(wait_by_service)


mentions_wait        False  True 
service_type                     
Citizen Services        24      0
Citizen services         1      0
Employment              19      5
Employment Services      1      0
Housing                 15      6
Transport               24      0
Transport                1      0
WASTE                    1      0
Waste                   24      0


### Compare waiting indicator and satisfaction


In [14]:
wait_score = (
    df.groupby("mentions_wait")["satisfaction_score"]
      .agg(["count", "mean", "median"])
)

display(wait_score.round(2))


               count  mean  median
mentions_wait                     
False            107  4.18    4.20
True              10  3.39    3.65


### Required written interpretation

Explain why this is a **descriptive association based on a simple coding rule**, not evidence that waiting causes satisfaction outcomes.

Also check how many waiting-related comments have a non-missing satisfaction score, so the number of comments matching the text rule and the number contributing to the score average are different.

---


## Example 3.14.8 — Write the Initial Data Exploration Report

[Open Example 3.14.8 on the website](https://asmrabbi.github.io/E26_TAN7_Scripting_CPH/#data-i/3.14/example-3.14.8)


Create a short report with these headings.

### A. Dataset and provenance

Include:

- dataset name;
- source;
- synthetic status;
- unit of observation.

### B. Structure

Include:

- number of rows/columns;
- key variables;
- conceptual types.

### C. Preliminary quantitative exploration

Include at least:

- one total;
- one mean;
- one median;
- one category percentage;
- one grouped comparison;
- one derived measure.

Every result must state what it counts/measures.

### D. Data-quality concerns

Include at least five identified issues and explain why they matter.

### E. Text/qualitative exploration

Include:

- manually observed ideas;
- at least one computational coding rule;
- match count;
- limitation of the rule.

### F. Three preliminary findings

Each finding should contain:

```text
WHAT WAS OBSERVED
+
HOW IT WAS CALCULATED/FOUND
+
WHAT LIMITS THE INTERPRETATION
```

### G. Next steps for Lecture 6

Identify which issues require:

- category standardisation;
- type conversion;
- date handling;
- missing-data decisions;
- duplicate verification;
- range validation;
- text preparation.

---

<details>
<summary>Open additional interpretation for Examples 3.14.1–3.14.8</summary>
